src/Xamarin.Android.Build.Tasks/Linker/MonoDroid.Tuner/FindTypeMapObjectsStep.cs

```diff
@@ -0,0 +1,87 @@
+#nullable enable
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Android.Build.Tasks;
+using Microsoft.Build.Utilities;
+using Mono.Cecil;
+using Mono.Linker.Steps;
+using Xamarin.Android.Tasks;
+
+namespace MonoDroid.Tuner;
+
+/// <summary>
+/// Scans an assembly for JLOs that need to be in the typemap and writes them to an XML file.
+/// </summary>
+public class FindTypeMapObjectsStep : BaseStep, IAssemblyModifierPipelineStep
+{
+	public bool Debug { get; set; }
+
+	public bool ErrorOnCustomJavaObject { get; set; }
+
+	public TaskLoggingHelper Log { get; set; }
+
+	public FindTypeMapObjectsStep (TaskLoggingHelper log) => Log = log;
+
+	public void ProcessAssembly (AssemblyDefinition assembly, StepContext context)
+	{
+		var destinationTypeMapXml = TypeMapObjectsXmlFile.GetTypeMapObjectsXmlFilePath (context.Destination.ItemSpec);
+
+		// We only care about assemblies that can contains JLOs
+		if (!context.IsAndroidAssembly) {
+			Log.LogDebugMessage ($"Skipping assembly '{assembly.Name.Name}' because it is not an Android assembly");
+			TypeMapObjectsXmlFile.WriteEmptyFile (destinationTypeMapXml, Log);
+			return;
+		}
+
+		var types = ScanForJavaTypes (assembly);
+
+		var xml = new TypeMapObjectsXmlFile {
+			AssemblyName = assembly.Name.Name,
+		};
+
+		if (Debug) {
+			var (javaToManaged, managedToJava) = TypeMapCecilAdapter.GetDebugNativeEntries (types, Context, out var foundJniNativeRegistration);
+
+			xml.JavaToManagedDebugEntries.AddRange (javaToManaged);
+			xml.ManagedToJavaDebugEntries.AddRange (managedToJava);
+			xml.FoundJniNativeRegistration = foundJniNativeRegistration;
+
+			if (!xml.HasDebugEntries) {
```

Comment:
I wonder if this block and the "equivalent" block in the `else` branch should be moved to the end…